From 55839af4ec927a9e47ee1ac838e02b1e3f879557 Mon Sep 17 00:00:00 2001
From: Muminul Islam <muislam@microsoft.com>
Date: Mon, 3 Jul 2023 17:56:15 -0700
Subject: [PATCH] Add Memory Hole to bitmap size

Signed-off-by: Muminul Islam <muislam@microsoft.com>
---
hypervisor/src/mshv/bitmap.rs | 9 +++++++--
1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/hypervisor/src/mshv/bitmap.rs b/hypervisor/src/mshv/bitmap.rs
index b2fa2c10..ee2b1d79 100644
--- a/hypervisor/src/mshv/bitmap.rs
+++ b/hypervisor/src/mshv/bitmap.rs
@@ -15,6 +15,7 @@ pub trait AtomicBitmapOps {
}

const INDEX_MASK:usize = 63;
+const ONE_GB: usize = 1024 * 1024 *1024;

impl AtomicBitmapOps for AtomicU64 {
     fn bit_size(&self) -> usize {
@@ -61,8 +62,12 @@ impl SimpleAtomicBitmap {
         }
     }
     pub fn new_with_bytes(size: usize, page_size: usize) -> Self {
-        let mut num_pages = size / page_size;
-        if size % page_size > 0 {
+        let mut new_size = size;
+        if size > 3 * ONE_GB {
+            new_size = size + ONE_GB;
+        }
+        let mut num_pages = new_size / page_size;
+        if new_size % page_size > 0 {
             num_pages += 1;
         }
         SimpleAtomicBitmap::new(num_pages)
--
2.34.1